cmd/juju/charmhub/infowriter.go

```diff
@@ -0,0 +1,275 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package charmhub
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/juju/charm/v7"
+	"github.com/juju/cmd"
+	"github.com/juju/errors"
+	"gopkg.in/yaml.v2"
+
+	"github.com/juju/juju/api/charmhub"
+	"github.com/juju/juju/cmd/output"
+)
+
+// Note:
+// Using yaml formatting for most of the juju info output,
+// to keep it similar to snap info, is easiest done in yaml.
+// There are exceptions, slices of strings and tables.  These
+// are transformed into strings.
+
+// InfoWriter defines what is needed to print info.
+type InfoWriter interface {
+	Print() error
+}
+
+func newInfoWriter(ctx *cmd.Context, in *charmhub.InfoResponse) InfoWriter {
+	iw := &infoWriter{
+		w:        ctx.Stdout,
+		warningf: ctx.Warningf,
+		in:       in,
+	}
+	if iw.in.Type == "charm" {
+		return &charmInfoWriter{infoWriter: iw}
+	}
+	return &bundleInfoWriter{infoWriter: iw}
```

Comment:
I bet `infoWriter` and the others don't need the pointers, if that's the case we should remove and rename `newInfoWriter` to `makeInfoWriter`.

Reply:
It's not, only the unmarshal methods required a pointer receiver.